Add tool to try Mapael

neveldo opened the pull request:
Hello @Indigo744 ,

Here is an improved version for the test page : https://www.vincentbroute.fr/mapael/try.html . Added : 

- A way to see the hovered area ID, pageX, pageY, map X and map Y.
- A button to export the map to JSFiddle
- A button to save the map into a file
- A way to upload a CSV file that contains areas/plots data to be displayed on the map

It can still be improved (add a way to chose Raphael / Mapael version, retrieve the maps from Github, etc), but I think that it's not bad for a first version. I hope you will like it :+1:)

Related issue : https://github.com/neveldo/mapael-documentation/issues/6#issuecomment-340720870

Indigo744 commented:
This is so cool! I love it :smile:

Some small remarks regarding the bottom information: 

1. Move `Hovered area ID` info at the beginning of string, and add a line break after it so the mouse coordinates are on a second line
2. When clicked on the map to fix the value, maybe add a reticle (I mean only drawing two lines that are crossing on the point). This could be done maybe with jQuery by simply adding two absolutely positioned element?
3. Add lat/long coordinates if they are available (check first if a call return something else than (0, 0) which would mean the function isn't implemented)
4. Duplicate these information also at top of the map, as for some big map it's hard to scroll down and up to check values... Or maybe add a `max-height` value to the map container so it's always visible (maybe use the `vh` unit, like `70vh` to be sure it's not larger than the viewport.

Indigo744 commented:
Also, the `export to JSFiddle` button generates a weird URL with mootools as dependency... e.g. http://jsfiddle.net/api/post/mootools/1.2/dependencies/more/

neveldo commented:
Glad you like it ! Here we go for the reticle : https://www.vincentbroute.fr/mapael/try.html . I have also fixed the form action for JSFiddle export.

However, how could we retrieve a latitude and longitude from (x,y) coordinates ? getCoords() is just a one-way function :(

Indigo744 commented:
Aww right... I didn't though about that... 😞

Indigo744 commented:
About the reticle: why should I select the `Spot a point on click` option? ~~It would be better to keep it simple: if I click (and not dragging), add the reticle.~~ It is actually good to avoid issue with onClick map event. However, dragging should be detected, right?

Indigo744 commented:
Also, I still think we should limit the height of the map to always have it inside the viewport. No?

neveldo commented:
> It would be better to keep it simple: if I click (and not dragging), add the reticle.

I added this checkbox having in mind the users who need to test regular click events on a map (for instance, area coloration on click, zoom on click, etc). It would be weird for them to have a reticle drawn on the map whenever they click on it !

> However, dragging should be detected, right?

Yes, of course !

> Also, I still think we should limit the height of the map to always have it inside the viewport. No?

Done ! I fixed the height and add an `overflow-y:auto;` to the container. It looks better, however, it seems that the scrollbar causes two bugs : 

- The map overflow the right boundary of the container
- The reticle is not well positionned

(you can see them with the maps of France for instance. Maybe I should be another solution without using a scrollbar ?

Indigo744 commented:
In Firefox, I don't even see the scrollbars actually:
![image](https://user-images.githubusercontent.com/7137528/32702444-af7556dc-c7e7-11e7-8b92-dee084cfd911.png)

Indigo744 commented on try.html:
```diff
@@ -156,35 +163,56 @@
+            <div class="mapExtraInfo">
+                Hovered area ID : <span class="area">-</span>
+                <br />pageX : <span class="pageX">0</span>, pageY : <span class="pageY">0</span> - X : <span class="x">0</span>, Y : <span class="y">0</span>
+                <br /><label for="spotPoint">Spot a point on click</label> <input type="checkbox" class="spotPoint" name="spotPoint" />
```
Add `id="spotPoint"` to checkbox input in order to register the click on label.

neveldo commented:
Strange, as I have only used regular CSS : 

```
            height:70vh;
            overflow-y:auto;
```

I will check it asap !

Indigo744 commented on try.html:
```diff
@@ -0,0 +1,766 @@
+                <div class="row">
+                    <div class="col-sm-4">
+                        <label for="mapSelector">Map</label>
+                        <select class="form-control" id="mapSelector" name="mapSelector"></select>
```
Clicking on the Map label seems to trigger the focus on the select, but it would be best if it opens it as well?

Indigo744 commented:
In another project, I had an image that needed to be responsive, but I wanted to apply a max-width *and* a max-height. I had it working at last using this [SO answer](https://stackoverflow.com/a/33420186/8861729). Don't know it it could apply here.

Indigo744 commented:
In the JSFiddle export, could the JS code for Mapael appear in the Javascript part? So people can easily modifiy it :smile:

neveldo commented:
It seems that I have not uploaded the last version, you should now see the scrollbar (and the issue on maps of France. It seems there is no problem on argentina map for instance, I need to investigate).

> Add id="spotPoint" to checkbox input in order to register the click on label.

Done

> Clicking on the Map label seems to trigger the focus on the select, but it would be best if it opens it as well?

Done

> In the JSFiddle export, could the JS code for Mapael appear in the Javascript part? So people can easily modifiy it :smile:

Unfortunately, no :( . There is a known bug with the POST feature on JSFiddle : the external libraries are loaded into random order and raise errors if the order is wrong. See https://stackoverflow.com/questions/41927911/post-to-jsfiddle-shows-external-resources-in-wrong-order

neveldo commented:
I have replaced the `overflow-y:auto;` by `overflow-y:scroll;` on the `.mapcontainer` div, and in seems to solve the width issue.

Indigo744 commented:
The issue highlighted by SO regards external resources. But can we simply keep the libraries in the HTML script tags, like now, but only put the init code in JS part (with one important thing: to select the `onDomready` load type in JS options)?

Like this:
![image](https://user-images.githubusercontent.com/7137528/32714586-dfd360aa-c84e-11e7-91d0-9c1eba53e472.png)

neveldo commented:
Hello @Indigo744 ,

Thanks, I didn't know about this option from JSFiddle, it should be ok now :)

Indigo744 commented:
Another option would have been to use the `$.ready()` of jQuery, but this one is easier :smile:

neveldo commented:
You are right, there is a little less code displayed in JSFiddle this way !

neveldo commented:
Hello @Indigo744 ,

Do you think that the tool is ok to be merged now, or do you have in mind some more improvements ?

Indigo744 commented:
Yes! It's a great tool 😃
